### ISSUE TITLE
[BUG?] Переключение между Англ-Рус-Укр

### ISSUE THREAD

**Issue by iG8R** [Krot66/LangBarXX#17]
Подскажите, пожалуйста, каким образом реализовать автоматическое переключение раскладки между Англ-Рус-Укр языками набора?
Например, раскладка изначально установлена на Англ, при попытке набора на украинском происходит переключение на русский и даже если дальше пытаться набирать украинские слова, все равно остается на прежней раскладке.

И возможно ли сделать так, чтобы клавиша "Pause" работала не только первые 2 раза, как в скринкасте, когда после слова "Невыдкладно!" клавиша "Pause" была нажата более 10 раз, но сработали только первые 2 и происходило циклическое переключение между Англ-Рус-Укр раскладками, как в PuntoSwitcher?

[LangBar-us-rus-ukr-switching-and-Pause-Video_2024-05-02_17-08-54.webm](https://github.com/Krot66/LangBarXX/assets/11407417/76e2a3f8-f93a-4499-a4e4-83f092bc0ca0)

<img src="https://github.com/Krot66/LangBarXX/assets/11407417/6a20790b-568d-42f6-8214-91dd4b3174a9" width="450">

<img src="https://github.com/Krot66/LangBarXX/assets/11407417/6ae0f527-a083-4332-9cf7-3b470ef43bb0" width="350">

**Comment by Krot66** [comment]
Дело в том, что если вы набираете слова на одном языке, можно сверять это слово только с парой словарей - текущего и какого-то другого. Можно на разных языках проверять один, как в случае режима одного языка. Сделать автоматическое переключение между трех языков сложно программно, требует больших задержек времени, да и недостаточно эффективно. Сочетания русский-белорусский работают, но не слишком четко из-за сходств языков.
Насчет циклических переключений Pause я не думал. Мне кажется, в программе достаточно других способов переключения раскладки (смотри Раскладки и флажки)

**Comment by iG8R** [comment]
Спасибо за столь быстрый ответ.
Эх, очень жаль, т.к. тогда теряется смысл перехода с PuntoSwitcher на LangBar++, поскольку и в том и другом случае все равно приходится вручную переключаться на третий язык, и в случае с Punto за это отвечает только одна кнопка - Pause, потому не надо запоминать другие сочетания, что получается немного удобнее + происходит циклическое переключение между раскладками.

PS. Когда у меня стояла Win7, была маленькая программка, которая отлично умела автопереключать раскладки анг-рус-укр-нем и делала она это крайне и крайне правильно, за редким исключением, к сожалению в Win10 она перестала работать. Прошло уже много времени и сейчас, к сожалению, не могу вспомнить как она называлась, но в описании помню было что-то схожее по алгоритму работы Punto, т.е. не на базе сверки по словарям, только автор его усовершенствовал.

**Comment by Krot66** [comment]
Ручное преобразование работает на любой язык - посмотрите справку. Программа автоматически и точно переключающая раскладку между 4 языками для меня есть нечто невозможное. Если вспомните, напишите

**Comment by iG8R** [comment]
В том-то и дело, вручную я и сейчас переключаю в Punto с помощью всего одной клавиши Pause на нужный мне язык, и тут вдруг увидел Вашу программу и подумал, что это можно осуществить автоматически.

> Программа автоматически и точно переключающая раскладку между 4 языками для меня есть нечто невозможное. Если вспомните, напишите

Да, хорошо.

**Comment by Krot66** [comment]
Я не понял: речь идет о циклическом **ручном** преобразовании последнего слова. Это сделать можно. Сейчас программа делает это по правый Ctrl+1, 2... Я как-то не думал над этим и не встречал аналогов

**Comment by iG8R** [comment]
Да, все верно, Pause в Punto осуществляет циклическое **ручное** преобразование последнего слова в те раскладки, которые установлены на данный момент на компе. Вот если бы еще можно было при этом исключать некоторые из них, было бы вообще замечательно. Например, в моей системе установлены следующие языки

![image](https://github.com/Krot66/LangBarXX/assets/11407417/ad154ca9-a91f-416b-8311-edf589eae837)

 И было бы замечательно, если бы из цикличного автоисправления был исключен Russian со знаками вопроса `??`.

Просто Ctrl+1,2... не так удобно, как просто нажимать всего одну клавишу Pause, которая практически ни в одной программе не используется, плюс у многих, в том числе и у меня, наверняка уже многие комбинации клавиш заняты другими программами, например, Ctrl+1,2..9,0 заняты под программу работы с буфером обмена - выбор сохраненных данных и вставка в место, где на данный момент расположена каретка. Shift+1,2..9,0 - тоже заняты для запуска скриптов в той же программе по работе с буфером обмена - ArsClip https://www.joejoesoft.com/vcms/97/

**Comment by Krot66** [comment]
Щелкните по языку правой кнопкой и переименуйте сокращение раскладки. Насчет исключений я понял

**Comment by iG8R** [comment]
> Щелкните по языку правой кнопкой и переименуйте сокращение раскладки.

Спасибо. Хотя по `??` мне и так понятно, с какой раскладкой я имею дело, просто мне этот язык (№3) вообще не нужен, т.к. он полностью дублирует язык №2, но удалить его из системы я не могу, т.к. на нем завязаны все региональные установки (формат времени, валюта и т.д.). Просто хотелось бы его исключить из циклического автопереключения.
Кстати, он потому и отображается в виде `??`, потому что я его уже однажды удалял, т.к. посчитал, что он не понадобится из-за наличия языка №2, но оказалось, что некоторые региональные установки из языка №2 не совсем мне подходят, потому пришлось вернуть, и уже после возвращения он стал так отображаться.

**Comment by Krot66** [comment]
Он выпадает из перечня кодировок [Microsoft Windows LCID List - Decimal and Hex - All locale codes IDs](https://www.ryadel.com/en/microsoft-windows-lcid-list-decimal-and-hex-all-locale-codes-ids/). Скорее всего, это модифицированная раскладка

**Comment by iG8R** [comment]
Это стандартная раскладка, просто она стала такой, когда я её восстановил через PowerShell после удаления.

Нашел программы, которые имеют автопереключение раскладок сразу на несколько языков, правда они уже не работают в Win10.
Сначала пользовался `Keyboard Ninja` - самая лучшая программа, затем при переходе на Win7 пришлось перейти на `KeyboardSwitcher` - как бы улучшенная версия, но на самом деле некоторые функции, что были у Ninja отсуствуют.

https://web.archive.org/web/20230529145611/http://keyboard-ninja.com/
https://web.archive.org/web/20230530102208/http://www.keyswitcher.com/

**Comment by Krot66** [comment]
Циклическое преобразование сделано в версии 1.7.8
